src/Build/Globbing/CompositeGlob.cs

```diff
@@ -5,33 +5,30 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
+using Microsoft.Build.Shared;
 
 namespace Microsoft.Build.Globbing
 {
     /// <summary>
-    ///     A Composite glob
+    ///     A composite glob that returns a match for an input if any of its
+    ///     inner globs match the input (disjunction).
     /// </summary>
     public class CompositeGlob : IMSBuildGlob
     {
+        private readonly ImmutableArray<IMSBuildGlob> _globs;
+
         /// <summary>
         /// The direct children of this composite
         /// </summary>
-        public IEnumerable<IMSBuildGlob> Globs { get; }
+        public IEnumerable<IMSBuildGlob> Globs => _globs;
 
         /// <summary>
         ///     Constructor
         /// </summary>
         /// <param name="globs">Children globs. Input gets shallow cloned</param>
         public CompositeGlob(IEnumerable<IMSBuildGlob> globs)
-        {
-            // ImmutableArray also does this check, but copied it here just in case they remove it
-            if (globs is ImmutableArray<IMSBuildGlob>)
-            {
-                Globs = (ImmutableArray<IMSBuildGlob>)globs;
-            }
-
-            Globs = globs.ToImmutableArray();
-        }
+            : this(globs is ImmutableArray<IMSBuildGlob> immutableGlobs ? immutableGlobs : globs.ToImmutableArray())
```

Comment:
ToImmutableArray already does that check for you: https://source.dot.net/#System.Collections.Immutable/System/Collections/Immutable/ImmutableArray.cs,7ec519665e45a686,references

Reply:
Thanks. The old code called this out but explained it as "in case they remove that". Then again the old code's attempt to optimise didn't work anyway. I'll file a follow up PR that removes it.

That said, there's no code in this repo that uses this overload any more. I would remove it, but it's public.